File: sections/47_meet_in_the_middle/problems/d_exact_sum_max_items/solution.py
import sys


def enumerate_states(values: list[int]) -> list[tuple[int, int]]:
    states = [(0, 0)]
    for value in values:
        states.extend((total + value, count + 1) for total, count in states[:])
    return states


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, target = data[0], data[1]
    values = data[2 : 2 + n]
    middle = n // 2
    left_states = enumerate_states(values[:middle])
    right_states = enumerate_states(values[middle:])

    best_right_count: dict[int, int] = {}
    for total, count in right_states:
        previous = best_right_count.get(total, -1)
        if count > previous:
            best_right_count[total] = count

    answer = -1
    for total, count in left_states:
        other_count = best_right_count.get(target - total)
        if other_count is not None:
            answer = max(answer, count + other_count)
    print(answer)


if __name__ == "__main__":
    main()

File: sections/47_meet_in_the_middle/problems/d_exact_sum_max_items/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using State = pair<long long, int>;

vector<State> enumerate_states(const vector<long long>& values) {
    vector<State> states{{0, 0}};
    states.reserve(1ULL << values.size());
    for (long long value : values) {
        int old_size = (int)states.size();
        for (int i = 0; i < old_size; i++) {
            states.push_back({states[i].first + value, states[i].second + 1});
        }
    }
    return states;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    long long target;
    if (!(cin >> n >> target)) return 0;
    vector<long long> values(n);
    for (long long& value : values) cin >> value;

    vector<long long> left_values(values.begin(), values.begin() + n / 2);
    vector<long long> right_values(values.begin() + n / 2, values.end());
    vector<State> left_states = enumerate_states(left_values);
    vector<State> right_states = enumerate_states(right_values);

    unordered_map<long long, int> best_right_count;
    best_right_count.reserve(right_states.size() * 2);
    for (const State& state : right_states) {
        best_right_count[state.first] = max(best_right_count[state.first], state.second);
    }

    int answer = -1;
    for (const State& state : left_states) {
        auto it = best_right_count.find(target - state.first);
        if (it != best_right_count.end()) answer = max(answer, state.second + it->second);
    }
    cout << answer << '\n';
    return 0;
}
